// qiita/lib/qiita.js
// Generated by CoffeeScript 1.7.1
(function() {
  var Qiita, Resource,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Resource = require("./resource");

  Qiita = (function(_super) {
    var DELETE, GET, PATCH, POST, PUT, RESOURCES;

    __extends(Qiita, _super);

    GET = "get";

    POST = "post";

    PUT = "put";

    DELETE = "delete";

    PATCH = "patch";

    RESOURCES = {
      comments: {
        path: "/comments/:id",
        methods: [GET, PATCH, DELETE],
        resource: {
          thank: {
            path: "/thank",
            methods: [PUT, DELETE]
          }
        }
      },
      items: {
        path: "/items/:item_id",
        methods: [GET, POST, PATCH, DELETE],
        resources: {
          comments: {
            path: "/comments",
            methods: [GET, POST]
          },
          stock: {
            path: "/stock",
            methods: [PUT, DELETE]
          }
        }
      },
      tags: {
        path: "/tags/:id",
        resources: {
          items: "/items"
        }
      },
      templates: {
        path: "/templates/:id",
        methods: [GET, POST, PATCH, DELETE]
      },
      projects: {
        path: "/projects/:id",
        methods: [GET, POST, PATCH, DELETE]
      },
      users: {
        path: "/users/:id",
        resources: {
          following_tags: "/following_tags",
          followees: "/followees",
          followers: "/followers",
          items: "/items",
          stocks: "/stocks"
        }
      },
      expanded_templates: "/expanded_templates",
      authenticated_user: "/authenticated_user"
    };

    function Qiita(config) {
      var host;
      if (config == null) {
        config = {};
      }
      host = config.team != null ? "" + config.team + ".qiita.com" : "qiita.com";
      Qiita.__super__.constructor.call(this, {
        token: config.token,
        path: "https://" + host + "/api/v2",
        resources: RESOURCES
      });
    }

    return Qiita;

  })(Resource);

  module.exports = Qiita;

}).call(this);
